src/main/java/net/spy/memcached/tls/TLSConnectionManager.java

```diff
@@ -0,0 +1,211 @@
+
+            // If we were interrupted, get out
+            if (Thread.currentThread().isInterrupted()) {
+                this.close();
+                throw new InterruptedException("Thread interrupted while completing SSL Handshake");
+            }
+
+
+        } catch (SSLException | InterruptedException e) {
+            throw new RuntimeException("Caught exception during SSL Handshake", e);
+        }
+        return true;
+    }
+
+    @Override
+    public void close() throws IOException {
+        sslEngine.setEnableSessionCreation(false);
+        sslEngine.closeOutbound();
+        sslEngine.closeInbound();
+    }
+
+    private void handleHandshakeWrapResult(SocketChannel socketChannel, SSLEngineResult result) throws SSLException {
+        switch (result.getStatus()) {
+            case BUFFER_UNDERFLOW:
+                // We should not get here
+                throw new SSLException("Buffer underflow occurred after wrap");
+            case BUFFER_OVERFLOW:
+                // If networkOutBuffer is too small to contain the response
+                networkOutBuffer = enlargeBuffer(networkOutBuffer, sslEngine.getSession().getPacketBufferSize());
+                // Try again with new larger buffer
+                break;
+            case OK:
+                sendNetworkBuffer(socketChannel);
+                break;
+            case CLOSED:
+                // We need to tell the server we're closing
+                sendNetworkBuffer(socketChannel);
+                // The next status will be NEED_UNWRAP and we'll need to pre-emptively clear the input network data
+                networkInBuffer.clear();
+                break;
+        }
+    }
+
+    private void handleHandshakeUnwrapResult(SocketChannel socketChannel, SSLEngineResult result) throws SSLException {
+        switch (result.getStatus()) {
+            case BUFFER_UNDERFLOW:
+                // If the network buffer is too small
+                networkInBuffer = enlargeBuffer(networkInBuffer, sslEngine.getSession().getPacketBufferSize());
+                break;
+            case BUFFER_OVERFLOW:
+                // if networkInBuffer is larger than appInBuffer
+                appInBuffer = enlargeBuffer(appInBuffer, sslEngine.getSession().getApplicationBufferSize());
+                // try again with larger buffer
+                break;
+            case OK:
+            case CLOSED:
+                break;
+        }
+    }
+
+    private static ByteBuffer enlargeBuffer(ByteBuffer buffer, int suggestedCapacity) {
+        if (suggestedCapacity > buffer.capacity()) {
+            return ByteBuffer.allocate(suggestedCapacity);
+        } else {
+            // If the suggested capacity is still too small, double the size
+            return ByteBuffer.allocate(buffer.capacity() * 2);
+        }
+    }
+
+    private void sendNetworkBuffer(SocketChannel socketChannel) throws SSLException {
+        networkOutBuffer.flip(); // Change from read to write
```

Comment:
So we don't need to flip this back because the handshake would be completed afterwards?

Reply:
that's my guess too